File: frontends/telegram/keyboards/reminders.py
"""Reminders keyboards for the Telegram bot."""

import calendar
from datetime import datetime, timezone, timedelta
from typing import Any, Dict, List

from telegram import InlineKeyboardButton, InlineKeyboardMarkup

from ..config import TIMEZONE_OFFSET_HOURS

_MONTH_NAMES = {
    1: "Январь",
    2: "Февраль",
    3: "Март",
    4: "Апрель",
    5: "Май",
    6: "Июнь",
    7: "Июль",
    8: "Август",
    9: "Сентябрь",
    10: "Октябрь",
    11: "Ноябрь",
    12: "Декабрь",
}

_SCHEDULE_TYPE_LABELS = {
    "daily": "Каждый день",
    "weekly": "По дням недели",
    "monthly": "Каждый месяц",
    "yearly": "Каждый год",
    "once": "Один раз",
    "custom_days": "Каждые N дней",
}


def get_reminders_list_keyboard(
    reminders: List[Dict[str, Any]],
    page: int = 0,
    per_page: int = 5,
) -> InlineKeyboardMarkup:
    total = len(reminders)
    total_pages = max(1, (total + per_page - 1) // per_page)
    page = max(0, min(page, total_pages - 1))
    start = page * per_page
    end = min(start + per_page, total)

    keyboard = []
    for r in reminders[start:end]:
        label = r["title"][:30]
        keyboard.append(
            [
                InlineKeyboardButton(f"🔔 {label}", callback_data="reminder:noop"),
                InlineKeyboardButton("🗑", callback_data=f"reminder:delete:{r['id']}"),
            ]
        )

    if total_pages > 1:
        nav: List[InlineKeyboardButton] = []
        if page > 0:
            nav.append(InlineKeyboardButton("◀", callback_data=f"reminder:page:{page - 1}"))
        nav.append(
            InlineKeyboardButton(f"{page + 1}/{total_pages}", callback_data="reminder:noop")
        )
        if page < total_pages - 1:
            nav.append(InlineKeyboardButton("▶", callback_data=f"reminder:page:{page + 1}"))
        keyboard.append(nav)

    keyboard.append(
        [InlineKeyboardButton("➕ Создать", callback_data="reminder:create")]
    )
    keyboard.append([InlineKeyboardButton("◀ Назад", callback_data="reminder:back")])
    return InlineKeyboardMarkup(keyboard)


def get_task_confirm_keyboard() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        [
            [
                InlineKeyboardButton("✅ Да, создавать задачу", callback_data="reminder:task_confirm:yes"),
                InlineKeyboardButton("❌ Нет", callback_data="reminder:task_confirm:no"),
            ],
            [InlineKeyboardButton("❌ Отмена", callback_data="reminder:cancel")],
        ]
    )


def get_schedule_type_keyboard() -> InlineKeyboardMarkup:
    keyboard = [
        [
            InlineKeyboardButton("Каждый день", callback_data="reminder:type:daily"),
            InlineKeyboardButton(
                "По дням недели", callback_data="reminder:type:weekly"
            ),
        ],
        [
            InlineKeyboardButton("Каждый месяц", callback_data="reminder:type:monthly"),
            InlineKeyboardButton("Каждый год", callback_data="reminder:type:yearly"),
        ],
        [
            InlineKeyboardButton("Один раз", callback_data="reminder:type:once"),
            InlineKeyboardButton(
                "Каждые N дней", callback_data="reminder:type:custom_days"
            ),
        ],
        [InlineKeyboardButton("❌ Отмена", callback_data="reminder:cancel")],
    ]
    return InlineKeyboardMarkup(keyboard)


def get_reminder_cancel_keyboard() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        [[InlineKeyboardButton("❌ Отмена", callback_data="reminder:cancel")]]
    )


def get_reminder_calendar_keyboard(
    year: int, month: int, context_name: str
) -> InlineKeyboardMarkup:
    """Calendar keyboard for picking a date in reminder flows.

    context_name: "once" | "yr" (yearly month+day) | "pp" (postpone)
    """
    local_now = datetime.now(timezone.utc) + timedelta(hours=TIMEZONE_OFFSET_HOURS)
    today = local_now.date()

    keyboard: list[list[InlineKeyboardButton]] = []

    # Header with navigation
    month_name = _MONTH_NAMES[month]
    keyboard.append(
        [
            InlineKeyboardButton(
                "◀", callback_data=f"reminder:cal:prev:{context_name}"
            ),
            InlineKeyboardButton(f"{month_name} {year}", callback_data="reminder:noop"),
            InlineKeyboardButton(
                "▶", callback_data=f"reminder:cal:next:{context_name}"
            ),
        ]
    )

    # Weekday headers
    keyboard.append(
        [
            InlineKeyboardButton(d, callback_data="reminder:noop")
            for d in ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Вс"]
        ]
    )

    import datetime as dt_module

    for week in calendar.monthcalendar(year, month):
        row: list[InlineKeyboardButton] = []
        for day in week:
            if day == 0:
                row.append(InlineKeyboardButton(" ", callback_data="reminder:noop"))
                continue
            cell_date = dt_module.date(year, month, day)
            if cell_date < today:
                row.append(InlineKeyboardButton(" ", callback_data="reminder:noop"))
            else:
                date_str = cell_date.strftime("%Y-%m-%d")
                label = f"[{day}]" if cell_date == today else str(day)
                row.append(
                    InlineKeyboardButton(
                        label,
                        callback_data=f"reminder:cal:select:{date_str}:{context_name}",
                    )
                )
        keyboard.append(row)

    keyboard.append(
        [
            InlineKeyboardButton(
                "📅 Сегодня", callback_data=f"reminder:cal:today:{context_name}"
            ),
            InlineKeyboardButton("❌ Отмена", callback_data="reminder:cancel"),
        ]
    )

    return InlineKeyboardMarkup(keyboard)

File: frontends/telegram/tgkeyboards/reminders.go
package tgkeyboards

import (
	"fmt"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"

	"notes-bot/frontends/telegram/clients"
)

const remindersPerPage = 5

// ReminderNotification builds the action keyboard attached to a fired reminder message.
func ReminderNotification(reminderID int64, createTask bool, todayDate string) tgbotapi.InlineKeyboardMarkup {
	doneCB := fmt.Sprintf("reminder:done:%d:0", reminderID)
	if createTask && todayDate != "" {
		doneCB = fmt.Sprintf("reminder:done:%d:1:%s", reminderID, todayDate)
	}
	return tgbotapi.NewInlineKeyboardMarkup(
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("✅ Принято", doneCB),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("+1 ч", fmt.Sprintf("reminder:postpone_hours:1:%d", reminderID)),
			tgbotapi.NewInlineKeyboardButtonData("+3 ч", fmt.Sprintf("reminder:postpone_hours:3:%d", reminderID)),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("+1 д", fmt.Sprintf("reminder:postpone:1:%d", reminderID)),
			tgbotapi.NewInlineKeyboardButtonData("+3 д", fmt.Sprintf("reminder:postpone:3:%d", reminderID)),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("📅 Выбрать дату", fmt.Sprintf("reminder:custom_date:%d", reminderID)),
		),
	)
}

func RemindersList(reminders []*clients.ReminderInfo, page int) tgbotapi.InlineKeyboardMarkup {
	total := len(reminders)
	totalPages := (total + remindersPerPage - 1) / remindersPerPage
	if totalPages == 0 {
		totalPages = 1
	}
	page = max(0, min(page, totalPages-1))
	start := page * remindersPerPage
	end := start + remindersPerPage
	if end > total {
		end = total
	}

	var rows [][]tgbotapi.InlineKeyboardButton
	for _, r := range reminders[start:end] {
		label := r.Title
		runes := []rune(label)
		if len(runes) > 30 {
			label = string(runes[:30])
		}
		rows = append(rows, tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("🔔 "+label, "reminder:noop"),
			tgbotapi.NewInlineKeyboardButtonData("🗑", fmt.Sprintf("reminder:delete:%d", r.ID)),
		))
	}

	if totalPages > 1 {
		var nav []tgbotapi.InlineKeyboardButton
		if page > 0 {
			nav = append(nav, tgbotapi.NewInlineKeyboardButtonData("◀", fmt.Sprintf("reminder:page:%d", page-1)))
		}
		nav = append(nav, tgbotapi.NewInlineKeyboardButtonData(fmt.Sprintf("%d/%d", page+1, totalPages), "reminder:noop"))
		if page < totalPages-1 {
			nav = append(nav, tgbotapi.NewInlineKeyboardButtonData("▶", fmt.Sprintf("reminder:page:%d", page+1)))
		}
		rows = append(rows, nav)
	}

	rows = append(rows, tgbotapi.NewInlineKeyboardRow(
		tgbotapi.NewInlineKeyboardButtonData("➕ Создать", "reminder:create"),
		tgbotapi.NewInlineKeyboardButtonData("✍️ Текстом", "reminder:create_nl"),
	))
	rows = append(rows, tgbotapi.NewInlineKeyboardRow(
		tgbotapi.NewInlineKeyboardButtonData("◀ Назад", "reminder:back"),
	))

	return tgbotapi.NewInlineKeyboardMarkup(rows...)
}

// NLReminderConfirm shows after the LLM parses a natural-language reminder.
func NLReminderConfirm() tgbotapi.InlineKeyboardMarkup {
	return tgbotapi.NewInlineKeyboardMarkup(
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("✅ Создать", "reminder:nl_confirm"),
			tgbotapi.NewInlineKeyboardButtonData("✏️ Вручную", "reminder:create"),
			tgbotapi.NewInlineKeyboardButtonData("❌ Отмена", "reminder:cancel"),
		),
	)
}

func TaskConfirm() tgbotapi.InlineKeyboardMarkup {
	return tgbotapi.NewInlineKeyboardMarkup(
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("✅ Да, создавать задачу", "reminder:task_confirm:yes"),
			tgbotapi.NewInlineKeyboardButtonData("❌ Нет", "reminder:task_confirm:no"),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("❌ Отмена", "reminder:cancel"),
		),
	)
}

func ScheduleType() tgbotapi.InlineKeyboardMarkup {
	return tgbotapi.NewInlineKeyboardMarkup(
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("Каждый день", "reminder:type:daily"),
			tgbotapi.NewInlineKeyboardButtonData("По дням недели", "reminder:type:weekly"),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("Каждый месяц", "reminder:type:monthly"),
			tgbotapi.NewInlineKeyboardButtonData("Каждый год", "reminder:type:yearly"),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("Один раз", "reminder:type:once"),
			tgbotapi.NewInlineKeyboardButtonData("Каждые N дней", "reminder:type:custom_days"),
		),
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("❌ Отмена", "reminder:cancel"),
		),
	)
}

func ReminderCancel() tgbotapi.InlineKeyboardMarkup {
	return tgbotapi.NewInlineKeyboardMarkup(
		tgbotapi.NewInlineKeyboardRow(
			tgbotapi.NewInlineKeyboardButtonData("❌ Отмена", "reminder:cancel"),
		),
	)
}

// ReminderCalendar builds a calendar for picking a date in reminder flows.
// contextName: "once" | "yr" (yearly) | "pp" (postpone)
func ReminderCalendar(year, month int, contextName string, tzOffsetHours int) tgbotapi.InlineKeyboardMarkup {
	tz := time.FixedZone("local", tzOffsetHours*3600)
	now := time.Now().In(tz)
	today := time.Date(now.Year(), now.Month(), now.Day(), 0, 0, 0, 0, tz)

	var rows [][]tgbotapi.InlineKeyboardButton

	rows = append(rows, tgbotapi.NewInlineKeyboardRow(
		tgbotapi.NewInlineKeyboardButtonData("◀", fmt.Sprintf("reminder:cal:prev:%s", contextName)),
		tgbotapi.NewInlineKeyboardButtonData(fmt.Sprintf("%s %d", monthNames[month], year), "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("▶", fmt.Sprintf("reminder:cal:next:%s", contextName)),
	))

	rows = append(rows, tgbotapi.NewInlineKeyboardRow(
		tgbotapi.NewInlineKeyboardButtonData("Пн", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Вт", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Ср", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Чт", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Пт", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Сб", "reminder:noop"),
		tgbotapi.NewInlineKeyboardButtonData("Вс", "reminder:noop"),
	))

	firstDay := time.Date(year, time.Month(month), 1, 0, 0, 0, 0, tz)
	startOffset := int(firstDay.Weekday()+6) % 7
	daysInMonth := time.Date(year, time.Month(month)+1, 0, 0, 0, 0, 0, tz).Day()

	day := 1
	for row := 0; row < 6 && day <= daysInMonth; row++ {
		var weekRow []tgbotapi.InlineKeyboardButton
		for col := 0; col < 7; col++ {
			if (row == 0 && col < startOffset) || day > daysInMonth {
				weekRow = append(weekRow, tgbotapi.NewInlineKeyboardButtonData(" ", "reminder:noop"))
			} else {
				cellDate := time.Date(year, time.Month(month), day, 0, 0, 0, 0, tz)
				if cellDate.Before(today) {
					weekRow = append(weekRow, tgbotapi.NewInlineKeyboardButtonData(" ", "reminder:noop"))
				} else {
					dateStr := cellDate.Format("2006-01-02")
					label := fmt.Sprintf("%d", day)
					if cellDate.Equal(today) {
						label = fmt.Sprintf("[%d]", day)
					}
					weekRow = append(weekRow, tgbotapi.NewInlineKeyboardButtonData(
						label,
						fmt.Sprintf("reminder:cal:select:%s:%s", dateStr, contextName),
					))
				}
				day++
			}
		}
		rows = append(rows, weekRow)
		if day > daysInMonth {
			break
		}
	}

	rows = append(rows, tgbotapi.NewInlineKeyboardRow(
		tgbotapi.NewInlineKeyboardButtonData("📅 Сегодня", fmt.Sprintf("reminder:cal:today:%s", contextName)),
		tgbotapi.NewInlineKeyboardButtonData("❌ Отмена", "reminder:cancel"),
	))

	return tgbotapi.NewInlineKeyboardMarkup(rows...)
}
